credentials/apps/edx_credentials_extensions/edly_credentials_app/api/serializers.py

```diff
@@ -0,0 +1,63 @@
+"""
+Serializers for edly_api.
+"""
+from django.contrib.sites.models import Site
+
+from rest_framework import serializers
+
+from credentials.apps.credentials.models import ProgramCertificate, Signatory
+from credentials.apps.edx_credentials_extensions.edly_credentials_app.api.fields import Base64ImageField
+
+
+class SignatorySerializer(serializers.ModelSerializer):
+    """
+    Serializer for Signatory.
+    """
+    image = Base64ImageField(max_length=None, use_url=True)
+
+    class Meta:
+        model = Signatory
+        fields = ['id', 'name', 'title', 'image']
+
+
+class ProgramCertificateConfigurationSerializer(serializers.ModelSerializer):
+    """
+    Serializer for Program Certificate Configuration.
+    """
+    site = serializers.SlugRelatedField(slug_field='domain', queryset=Site.objects.all())
+    signatories = SignatorySerializer(many=True)
+
+    class Meta:
+        model = ProgramCertificate
+        fields = [
+            'site', 'is_active', 'signatories', 'program_uuid', 'use_org_name',
+            'include_hours_of_effort', 'language',
+        ]
+
+    def create(self, validated_data):
```

Comment:
By default nested serializers are read-only. We would have to create `create` & `update` methods to handle that manually.
https://www.django-rest-framework.org/api-guide/relations/#writable-nested-serializers

Reply:
@iam-mhaseeb that is what I'm suggesting. You should not use nested serializers if you need to perform add/update on entities instead use relational fields.

Reply:
Relational fields would point to single fields in serializers. I would explore custom relation fields to handle JSON representation like we want to use.
```
{
    "site": "edx.devstack.lms:18150",
    "is_active": true,
    "signatories": [
        {
            "name": "Jhon Doe",
            "title": "Director Certificates",
            "image": "data:image/png;base64,VBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII="
        }
    ],
    "program_uuid": "242681e5-b746-4204-8601-f0512a23b6fb",
    "use_org_name": true,
    "include_hours_of_effort": true,
    "language": "en"
}
```
https://www.django-rest-framework.org/api-guide/relations/#custom-relational-fields

Reply:
@iam-mhaseeb I think this is what @ziafazal is suggesting (https://github.com/edx/course-discovery/pull/1781/files):

<img width="1105" alt="Screen Shot 2020-12-02 at 4 28 36 PM" src="https://user-images.githubusercontent.com/5072991/100867207-ad8a9000-34bb-11eb-839b-d0760885fbbb.png">

Reply:
@ziafazal @zubair-arbi I tried 2 approaches to achieve this.

**1. Creating a custom field like below.**
 ```
class SignatoryRelatedField(serializers.RelatedField):
    queryset = Signatory.objects.all()

    def to_representation(self, instance):
        return json.dumps({
            'id': instance.id,
            'name': instance.name,
            'title': instance.title,
            'image': instance.image.url,
        })

    def to_internal_value(self, data):
        signatories = []
        for signatory in data.get('signatories', []):
            signatory_obj, _ = Signatory.objects.get_or_create(
                pk=signatory.get('id'),
                defaults=signatory
            )
            signatories.append(signatory_obj)

        return signatories
```
I believe it will help us in omitting custom create & update methods but the only problem with this approach is that `to_internal_value` restricts us in using `list` or `dict` as the return value. In our data representation, we can expect multiple signatories while creating a program certificate configuration as show in sample data below.
```
{
    "site": "edx.devstack.lms:18150",
    "is_active": true,
    "signatories": [
        {
            "name": "Jhon Doe",
            "title": "Director Certificates",
            "image": "data:image/png;base64,VBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII="
        }
    ],
    "program_uuid": "242681e5-b746-4204-8601-f0512a23b6fb",
    "use_org_name": true,
    "include_hours_of_effort": true,
    "language": "en"
}
```
**2. Defining related field for each field of Signatory in `ProgramCertificateConfigurationSerializer`**
```
class ProgramCertificateConfigurationSerializer(serializers.ModelSerializer):
    """
    Serializer for Program Certificate Configuration.
    """
    site = serializers.SlugRelatedField(slug_field='domain', queryset=Site.objects.all())
    image = serializers.SlugRelatedField(slug_field='image', queryset=Signatory.objects.all())
    name = serializers.SlugRelatedField(slug_field='name', queryset=Signatory.objects.all())
    title = serializers.SlugRelatedField(slug_field='title', queryset=Signatory.objects.all())
``` 
That way we can't send JSON request and we would have to send a multipart request and attach an image with that also it would represent only 1 Signatory and still, we would have to handle that data in custom `create & update` methods.

So, far I've seen related fields are used when we wanted to make ready only relation on 1 field of. a model or we wanted to create or update the model using 1 field. For example, Course can have a user as a related field containing a username as a slug field pointing to username for model object creation, update, or retrieval.
Even in the above example shared by @zubair-arbi they are telling mapping to single field `slug` of `subject` and they still had to use custom serializer while returning data as they can't return list or dict.

Reply:
@iam-mhaseeb Thanks for doing a detail discovery of different options. I think we don't have any other option than using nested serializers(as you did) and define `create` and `update` methods in them.